src/bson_mapper/utility.hpp

```diff
@@ -0,0 +1,143 @@
+ * into
+ * the bson_ostream callback?
+ * @tparam T   A type that is serializable to BSON using a BSONArchiver.
+ * @param  obj A serializable object
+ * @return     A BSON document value representing the given object.
+ */
+template <class T>
+bsoncxx::document::value to_document(const T& obj) {
+    bsoncxx::stdx::optional<bsoncxx::document::value> doc;
+    bson_mapper::bson_ostream bos([&doc](bsoncxx::document::value v) { doc = std::move(v); });
+    // TODO cereal::BSONOutputArchiver archiver(bos);
+    out_archiver_mock<T> archive(bos);
+    archive(obj);
+    return doc.value();
+}
+
+/**
+* Converts a bsoncxx document view to an object of the templated type through deserialization.
+* The object must be default-constructible.
+*
+* @tparam T A default-constructible type that is serializable using a BSONArchiver
+* @param v A BSON document view. If the BSON document does not match the schema of type T,
+* BSONArchiver will throw a corresponding exception.
+* TODO what does BSONArchiver throw?
+* @return Returns by value an object that corresponds to the given document view.
+*/
+template <class T>
+T to_obj(bsoncxx::document::view v) {
+    static_assert(std::is_default_constructible<T>::value,
+                  "Template type must be default constructible");
+    bson_mapper::bson_istream bis(v);
+    // TODO cereal::BSONInputArchive archive(bis);
+    in_archiver_mock<T> archive(bis);
+    T obj;
+    archive(obj);
+    return obj;
+}
+
+/**
+ * Fills a serializable object 'obj' with data from a BSON document view.
+ *
+ * @tparam T a type that is serializable using a BSONArchiver
+ * @param v A BSON document view. If the BSON document does not match the schema of type T,
+ * BSONArchiver will throw a corresponding exception.
+ * TODO what does BSONArchiver throw?
+ * @param obj A reference to a serializable object that will be filled with data from the gievn
+ * document.
+ */
+template <class T>
+void to_obj(bsoncxx::document::view v, T& obj) {
+    bson_mapper::bson_istream bis(v);
+    // TODO cereal::BSONInputArchive archive(bis);
+    in_archiver_mock<T> archive(bis);
+    archive(obj);
+}
+
+/*
+* This function converts an stdx::optional containing a BSON document value into an
+* stdx::optional
+* containing a deserialized object of the templated type.
+* @tparam T A type that is serializable to BSON using a BSONArchiver.
+* @param opt An optional BSON document view
+* @return   An optional object of type T
+*/
+template <class T>
+bsoncxx::stdx::optional<T> to_optional_obj(
+    const bsoncxx::stdx::optional<bsoncxx::document::value>& opt) {
+    if (opt) {
+        T obj = to_obj<T>(opt.value().view());
+        return {obj};
+    } else {
+        return {};
+    }
+}
+
+/**
+ * An iterator that wraps another iterator of serializable objects, and yields BSON document
+ * views
+ * corresponding to those documents.
+ *
+ * TODO what to do if serialization fails?
```

Comment:
The BSONOutputArchive will throw an exception which should propagate up through this. Right now, serialization failure should only occur if objects are missing names or if a class containing bsoncxx view types isn't wrapped in a class inheriting from UnderlyingBSONDataBase.

Both of these things are a user's fault and they need to be fixed for serialization to work properly so we shouldn't have to handle them here.

I guess serialization might also fail if it can't write to the stream but— @Machyne and @hanumantmk correct me if I'm wrong—that shouldn't happen if our streams are based on the bson_streambuf.